packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js

```diff
@@ -0,0 +1,388 @@
+) => {
+  poolValues[tokenIndexFrom] -= inputTokenValue;
+  poolValues[tokenIndexTo] += outputTokenValue;
+  return poolValues;
+};
+
+/**
+ * Computes the Stable Swap invariant (D).
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ * passed from the contract.
+ * @returns {bigint} d - the current price, in value form
+ *
+ */
+export const getD = poolValues => {
+  const N_COINS = poolValues.length;
+  // sum_x - Sum of all poolValues.
+  const sum_x = poolValues.reduce((prev, cur) => prev + cur, 0n);
+  if (sum_x === 0n) {
+    return 0n;
+  }
+  let d_prev = 0n;
+  let d = sum_x;
+  const Ann = A * N_COINS * N_COINS;
+  for (let i = 0; i < maxLoopLimit; i++) {
+    let dp = d;
+    // prod - product of all poolvalues
+    // dp = D^(n+1)/n^n(prod)
+    for (let j = 0; j < N_COINS; j++) {
+      dp = (dp * d) / (poolValues[j] * Nat(N_COINS));
+    }
+    d_prev = d;
+    // Non simplified form
+    // d = d-(dp+d(Ann-1)-Ann*sum_x)/(((dp*(n+1))/d)+(Ann-1))
+    d =
+      d -
+      (dp + d * (Nat(Ann) - 1n) - Nat(Ann) * sum_x) /
+        ((dp * (Nat(N_COINS) + 1n)) / d + (Nat(Ann) - 1n));
+    // Checks whether the iteration result is to the accuracy
+    // or one more iteration is required.
+    if (within10(d, d_prev)) {
+      return d;
+    }
+  }
+  return d;
+};
+
+/**
+ * Compute the swap amount `y` in proportion to `x`.
+ *
+ * @param {bigint} x - index of amount of inputReserve
+ * in the reserves array.
+ * @param {number} tokenIndexFrom - index of amount of inputReserve
+ * in the reserves array.
+ * @param {number} tokenIndexTo - index of amount of outputReserve
+ * in the reserves array.
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ * passed from the contract.
+ * @returns {bigint} y - the amount of swap out asset to be returned
+ * in exchange for amount x of swap in asset.
+ */
+
+export const getY = (x, tokenIndexFrom, tokenIndexTo, poolValues) => {
+  const d = getD(poolValues);
+  const N_COINS = poolValues.length;
+  const Ann = A * N_COINS * N_COINS;
+  let c = d;
+  let s = 0n;
+  let _x;
+  let xi;
+  // sum` - is sum of all pool values apart from the
+  // the swap out token's pool value.
+  // prod` - is the product of all pool values apart
+  // from the swap out token's poolValue.
+  // s = sum`
+  // c=(D^(n+1))/(n^n)*prod`
+  for (let i = 0; i < N_COINS; i++) {
+    if (i === tokenIndexFrom) {
+      _x = x;
+      xi = x;
+    } else if (i !== tokenIndexTo) {
+      _x = poolValues[i];
+      xi = poolValues[i];
+    } else {
+      _x = 0n;
+      xi = 1n;
+    }
+    s += _x;
+    c = (c * d) / (xi * Nat(N_COINS));
+  }
+  let y_prev = 0n;
+  let y = d;
+  for (let i = 0; i < maxLoopLimit; i++) {
+    y_prev = y;
+    // yi+1=yi-((Ann*y^2)+Ann*s*y-d*y*(Ann-1)-c)/(2Ann*y+Ann*s+D(Ann-1))
+    y =
+      y -
+      (Nat(Ann) * (y * y) + Nat(Ann) * s * y - d * y * Nat(Ann - 1) - c) /
+        (Nat(2 * Ann) * y + Nat(Ann) * s - d * Nat(Ann - 1));
```

Comment:
Here, we're only dividing once per iteration, but if we convert y to Nat each time, we're losing precision each step, while trying to use Newton's method to reach a very precise result. It would be better to use Ratio to keep numerator and denominator distinct. 

I think the withinEpsilon check should probably compare y_prev and y as floats and see if they're within .001% or something.

As BigInts, `within10` means something very different if the pools are in the thousands than if they're in the billions.  And the interesting case is when one of the pool balances gets driven to near zero.